File: plugins/orchestrator/skills/memory-bank/scripts/init_memories.py

import os
import shutil
import json
import urllib.request
import datetime

def get_utc_time():
    try:
        with urllib.request.urlopen("http://worldtimeapi.org/api/timezone/UTC", timeout=5) as response:
            data = json.loads(response.read().decode())
            return data["datetime"]
    except Exception as e:
        print(f"Warning: Could not fetch time from worldtimeapi.org ({e}). Using local UTC time.")
        return datetime.datetime.now(datetime.timezone.utc).isoformat()

def main():
    root_dir = os.getcwd()
    memories_dir = os.path.join(root_dir, ".memories")
    
    # Locate templates relative to this script
    script_dir = os.path.dirname(os.path.abspath(__file__))
    templates_dir = os.path.join(script_dir, "..", "assets", "templates")
    
    if not os.path.exists(memories_dir):
        print(f"Creating .memories directory at {memories_dir}")
        os.makedirs(memories_dir)
    else:
        print(f".memories directory already exists at {memories_dir}")

    # Initialize metadata.json if not exists
    metadata_path = os.path.join(memories_dir, "metadata.json")
    if not os.path.exists(metadata_path):
        initial_metadata = {"nodes": [], "edges": []}
        with open(metadata_path, 'w', encoding='utf-8') as f:
            json.dump(initial_metadata, f, indent=2)
        print("Created metadata.json")

    # Copy templates
    utc_time = get_utc_time()
    print(f"Using UTC time: {utc_time}")

    core_files = [
        "projectbrief.md",
        "productContext.md",
        "activeContext.md",
        "systemPatterns.md",
        "techContext.md",
        "progress.md",
        "_index.md"
    ]

    for filename in core_files:
        dest_path = os.path.join(memories_dir, filename)
        if not os.path.exists(dest_path):
            src_path = os.path.join(templates_dir, filename)
            if os.path.exists(src_path):
                with open(src_path, 'r', encoding='utf-8') as f:
                    content = f.read()
                
                # Replace placeholder
                content = content.replace("{{UTC_DATETIME}}", utc_time)
                
                with open(dest_path, 'w', encoding='utf-8') as f:
                    f.write(content)
                print(f"Created {filename}")
            else:
                print(f"Warning: Template {filename} not found at {src_path}")
        else:
            print(f"File {filename} already exists. Skipping.")

    print("\nMemory Bank initialized successfully.")
    print("Don't forget to run 'sync_metadata.py' to update the graph.")

if __name__ == "__main__":
    main()

File: plugins/orchestrator/skills/memory-bank/scripts/init_memories.ts
import { mkdirSync } from 'node:fs';
import { join, dirname } from 'node:path';
import { fileURLToPath } from 'node:url';

interface TimeApiResponse {
  datetime: string;
}

interface Metadata {
  nodes: unknown[];
  edges: unknown[];
}

async function getUtcTime(): Promise<string> {
  try {
    const response = await fetch('http://worldtimeapi.org/api/timezone/UTC', {
      signal: AbortSignal.timeout(5000),
    });
    const data: TimeApiResponse = await response.json();
    return data.datetime;
  } catch (error) {
    console.warn(`Warning: Could not fetch time from worldtimeapi.org (${error}). Using local UTC time.`);
    return new Date().toISOString();
  }
}

async function main(): Promise<void> {
  const rootDir = process.cwd();
  const memoriesDir = join(rootDir, '.memories');
  
  // Locate templates relative to this script
  const scriptDir = dirname(fileURLToPath(import.meta.url));
  const templatesDir = join(scriptDir, '..', 'assets', 'templates');
  
  // Create .memories directory if it doesn't exist
  try {
    mkdirSync(memoriesDir, { recursive: true });
    console.log(`Creating .memories directory at ${memoriesDir}`);
  } catch {
    console.log(`.memories directory already exists at ${memoriesDir}`);
  }
  
  // Initialize metadata.json if not exists
  const metadataPath = join(memoriesDir, 'metadata.json');
  try {
    const metadataFile = Bun.file(metadataPath);
    if (!(await metadataFile.exists())) {
      const initialMetadata: Metadata = { nodes: [], edges: [] };
      await Bun.write(metadataPath, JSON.stringify(initialMetadata, null, 2));
      console.log('Created metadata.json');
    }
  } catch {
    console.log('metadata.json already exists');
  }
  
  // Copy templates
  const utcTime = await getUtcTime();
  console.log(`Using UTC time: ${utcTime}`);
  
  const coreFiles = [
    'projectbrief.md',
    'productContext.md',
    'activeContext.md',
    'systemPatterns.md',
    'techContext.md',
    'progress.md',
    '_index.md',
  ];
  
  for (const filename of coreFiles) {
    const destPath = join(memoriesDir, filename);
    const destFile = Bun.file(destPath);
    
    if (!(await destFile.exists())) {
      const srcPath = join(templatesDir, filename);
      const srcFile = Bun.file(srcPath);
      
      if (await srcFile.exists()) {
        let content = await srcFile.text();
        
        // Replace placeholder
        content = content.replace('{{UTC_DATETIME}}', utcTime);
        
        await Bun.write(destPath, content);
        console.log(`Created ${filename}`);
      } else {
        console.log(`Warning: Template ${filename} not found at ${srcPath}`);
      }
    } else {
      console.log(`File ${filename} already exists. Skipping.`);
    }
  }
  
  console.log('\nMemory Bank initialized successfully.');
  console.log("Don't forget to run 'sync_metadata.ts' to update the graph.");
}

main().catch((error) => {
  console.error('Error:', error);
  process.exit(1);
});